Implement ECC APIs on Android using Android-built-in Crypto APIs

jkoritzinsky opened the pull request:
Implement the ECC key management, import/export, ECDSA, and ECDH APIs using the Java-based Crypto APIs available on Android.

cc: @bartonjs

jkoritzinsky committed:
Copy over OpenSSL ECC impl and start reimplementing with JNI. Currently implemented CryptoNative_EcKeyCreateByExplicitParameters.

jkoritzinsky committed:
Implement pal_eckey APIs and stub out required JNI interfaces. Error handling still todo.

jkoritzinsky committed:
Fill in implementations of ECC export methods.

jkoritzinsky committed:
Fix C build (linking still tbd).

jkoritzinsky committed:
Fix linking and initialize JNI.

jkoritzinsky committed:
Handle lifetime and exceptions.

jkoritzinsky committed:
Implement ECDSA APIs.

jkoritzinsky committed:
Add some needed exception handling and fix incorrect member name.

jkoritzinsky committed:
Merge branch 'master' into android-ecc

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_evp_cipher.c:
```diff
@@ -107,7 +107,7 @@ static jobject GetAlgorithmName(JNIEnv* env, intptr_t type)
     //if (type == CryptoNative_EvpAes256Cfb128()) return JSTRING("AES/CFB/NoPadding");
 
-    LOG_ERROR("This algorithm (%ld) is not supported", type);
+    LOG_ERROR("This algorithm (%ld) is not supported", (long)type);
```
This failed to build correctly on 32-bit Android without updates.

filipnavara commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+    {
+        jobject privateKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPrivateMethod);
+
+        jobject dBn = (*env)->CallObjectMethod(env, privateKey, g_ECPrivateKeyGetS);
```
What will happen if I call `ExportParameters(includePrivate: true)` on object that was constructed with public key only?

bartonjs replied:
This question, to me, is "how do Java exceptions work from the JNI call?"

akoeplinger replied:
Any Java method that can throw an exception needs to be check with `(*env)->ExceptionOccurred(env)`, see http://journals.ecs.soton.ac.uk/java/tutorial/native1.1/implementing/error.html

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.h:
```diff
@@ -0,0 +1,63 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#pragma once
```
Is this used in other places in the codebase? It is technically non-standard even though many compilers support it. We should confirm we are permitted to use this in this area.

elinor-fung replied:
A lot of the files in `src/libraries/Native` currently use `#pragma once`, including the ones already in `System.Security.Cryptography.Native.Android`

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.h:
```diff
@@ -0,0 +1,63 @@
+#include "pal_compiler.h"
+#include "pal_jni.h"
+
+typedef struct EC_KEY
```
Typical declarations here are to add an `_` prefix - `typedef struct _EC_KEY`. Again, just want to make sure we follow the current style.

jkoritzinsky replied:
Current style in the Android crypto PAL is to use the same name in the struct name and typedef. See the `pal_rsa.h` header.

AaronRobinsonMSFT replied:
I don't see any `struct`s declared in [`pal_rsa.h`](https://github.com/dotnet/runtime/blob/master/src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_rsa.h). But I see many matches for "typedef struct _" when I search the code base.

jkoritzinsky replied:
Your link points to the Linux (not Android) `pal_rsa.h`. This is the example I was thinking of:
https://github.com/dotnet/runtime/blob/0026a202ab4087d237f6be8824dff322d3a7d6be/src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_rsa.h#L15

AaronRobinsonMSFT replied:
Okay. Thanks.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.h:
```diff
@@ -0,0 +1,29 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Do we need an include guard here?

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c:
```diff
@@ -0,0 +1,112 @@
+static jobject
+CryptoNative_GetEsDsaSignatureObject()
+{
+    JNIEnv* env = GetJNIEnv();
```
Pass in `JNIEnv*`? It is done prior to all invocations of this function.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c:
```diff
@@ -0,0 +1,112 @@
+    jstring algorithmName = JSTRING("SHA1withECDSA");
+    jobject signatureObject = (*env)->CallStaticObjectMethod(env, g_SignatureClass, g_SignatureGetInstance, algorithmName);
+    (*env)->DeleteLocalRef(algorithmName);
+    if (CheckJNIExceptions()) return NULL;
```
Are we not printing the exception here? Does this indicate a sort of catastrophic failure?

jkoritzinsky replied:
The `CheckJNIExceptions()` method prints out the exceptions to the Android logs.

AaronRobinsonMSFT replied:
Okay. So `NULL` is okay here then, but we immediately return in all places where `NULL` is returned but don't record in that scenario. It seems to me that this code should be as boiler plate as possible and this function handles errors but does the caller so I am just trying to understand how reconciles ownership of when to report and when not to. Also, I think we can pass in `env` right?

jkoritzinsky replied:
I like having the `CheckJNIExceptions()` call as close to the throwing Java method call as possible for clarity. Since this method calls the Java `Signature.getInstance` method, it checks if it threw an exception.

The callers of this method don't need to check if it threw an exception since we already checked that.

I think a good way to encapsulate the model for JNI exception checking in the PAL is, "a JNI exception will never be left pending across a PAL function return." Or at least that's my mental model of it.

AaronRobinsonMSFT replied:
I think what I'm getting at is this function seems like 3 lines in 2 places and can be removed entirely.  Just my opinion. Up to you.

jkoritzinsky replied:
Yeah that's true. I'll think about removing it. We might want something similar for RSA signing, so I might generalize it for that.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,133 @@
+
+    jobject curveParameters = (*env)->CallObjectMethod(env, keySpec, g_ECPublicKeySpecGetParams);
+    
+
```
Extra line.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.h:
```diff
@@ -0,0 +1,74 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
```
Include guard?

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+    jobject* qy, int32_t* cbQy,
+    jobject* d, int32_t* cbD)
+{
+    assert(qx != NULL);
```
What is the logic to assert here as opposed to return an error code? In other functions it seems we do a hard check instead of simply asserting? Perhaps this is an internal function? I would recommend we be consistent here.

AaronRobinsonMSFT replied:
We guard in the error case for `d` but not here?

EgorBo replied:
I guess the asserts are copied from OpenSSL's CryptoNative_GetECKeyParameters

bartonjs replied:
The shims are considered 100% as internal functions.  We're inconsistent of asserting vs error checking (largely, IIRC, error checking was first, when we had the possibility of torn state of the native shim package and the caller package... now that we have shared frameworks some of the newer things assert)

AaronRobinsonMSFT replied:
> The shims are considered 100% as internal functions.
> now that we have shared frameworks some of the newer things assert

W00t. If these functions are internal only I would prefer to limit this sort of validation to asserts only. My assumption is we will always call these functions correctly and find incorrect internal usage with testing.

jkoritzinsky replied:
I've noticed that a lot of places in the crypto tests push the validation down to the shim/native library, so we can't use asserts for those ones since they'll take down the process.

bartonjs replied:
It all comes back to the current "time-saving" effort of reusing the OpenSSL-based implementations.  The tests (generally) don't care where things fail, just that CryptographicException-derived exceptions happen.  If you make ECDsaAndroid and it does more work in managed code you can do more asserting and less runtime error checking.

If there are tests that care about specifics for the exceptions I believe those should only care about using the concrete public ECDsaOpenSsl type, which should NOT work on Android since it can't speak pointers in an actually interoperable way (and the pointer interop is the only reason the type is public).

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+    const EC_KEY* key,
+    int32_t includePrivate,
+    ECCurveType* curveType,
+    jobject* qx, int32_t* cbQx,
```
Might it make sense to create a blittable in/out value type for these large pointer sets? I assume all these functions are for internal usage and not external consumption.

bartonjs replied:
My answer to this, and everything else in the shim API, is "do whatever makes sense for this interop".  Just make a new ECDsaAndroid/ECDiffieHellmanAndroid as internal types in System.Security.Cryptography.Algorithms and own the interop for them entirely.  There's no reason to stay aligned with the OpenSSL shim API... especially since I've already started the process of rewriting that one :smile:.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+
+            if ((p[0] & test) == test)
+            {
+                // TODO: test for overflow.
```
should we do this now?

EgorBo commented:
My two cents (unrelated to this PR)
1) I'd rewrite C to C++ honestly, it'd be less verbose e.g. `(*env)->Foo(env)` to `env->Foo()` at a cost of a minor size regression.
2) I'd make the declarations of classes/methods to be table-based on macros, e.g. something like this: https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/instrsxarch.h

EgorBo commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+    if (bBn) (*env)->DeleteLocalRef(env, bBn);
+    if (orderBn) (*env)->DeleteLocalRef(env, orderBn);
+    if (cofactorBn) (*env)->DeleteLocalRef(env, cofactorBn);
+    if (seedBn) (*env)->DeleteLocalRef(env, seedBn);
```
For simplicity a sort of a `ReleaseLReg` can be added (with null check inside)

AaronRobinsonMSFT commented:
> 1. I'd rewrite C to C++ honestly, it'd be less verbose e.g. `(*env)->Foo(env)` to `env->Foo()` at a cost of a minor size regression.

I don't think that is an option here - at least I was under the impression we were required to use pure C.

> 2. I'd make the declarations of classes/methods to be table-based on macros, e.g. something like this: https://github.com/dotnet/runtime/blob/master/src/coreclr/jit/instrsxarch.h

I think we can consider that pattern if we are keeping with C. However, if we are going with C++ that sort of macro magic should be avoided as it can't be debugged and modern C++ has mechanisms to avoid those patterns.

EgorBo commented:
> I don't think that is an option here - at least I was under the impression we were required to use pure C.

AFAIK, System.Native.* libs were initially C++ in corefx but then were rewritten to be: 1) Mono-friendly (Mono used to be C-only and couldn't consume/build C++ files) 2) to save some space (not much) by dropping stdlib.

AaronRobinsonMSFT commented:
>1) Mono-friendly (Mono used to be C-only and couldn't consume/build C++ files)

Yep. That is what I was alluding to. Does Mono now permit C++ in their scenarios?

EgorBo commented:
> > 1. Mono-friendly (Mono used to be C-only and couldn't consume/build C++ files)
> 
> Yep. That is what I was alluding to. Does Mono now permit C++ in their scenarios?

Well, this lib is a standalone dynamic lib so it doesn't matter at all - cmake will just build a `*.so` and Mono will be able to pinvoke it.

AaronRobinsonMSFT commented:
>Well, this lib is a standalone dynamic lib so it doesn't matter at all - cmake will just build a *.so and Mono will be able to pinvoke it.

I am fine with that if it is okay. We will have to make sure the target platforms provide the C++ runtime and ensure no exceptions cross the binary boundary but other than those two considerations I am all for it.

filipnavara commented:
There were some other reasons for the C++ > C conversion. Maybe @am11 will remember them? It may not matter in this case if it's specific to Android platform.

AaronRobinsonMSFT commented:
@filipnavara @am11 @VSadov Was the decision based on the Single File + self-contained work?

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,547 @@
+        // Get the reduction polynomial p
+        pBn = (*env)->CallObjectMethod(env, field, g_ECFieldF2mGetReductionPolynomial);
+    }
+    else
```
Are there no twisted edwards curves? (Curve25519 being the most popular one)

jkoritzinsky replied:
I don't believe there's an API to explicitly create them in the built-in Java support. In addition, the built-in provider doesn't seem to support Curve25519. The BouncyCastle provider supports it, so I think people just end up using that if they're using Java and they need this curve.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c:
```diff
@@ -0,0 +1,112 @@
+    // As a result, we see that there are 6 additional bytes in the DER encoding than the lengths of R and S combined.
+    // As the ECDSA algorithm is defined, the maximum length of R and S each is the bitlength of the order, so as a result
+    // we get the maximum size as 2 * bitlength of the order + 6.
+    const int derEncodingBytes = 6;
```
The DER encoding overhead is not a constant.

Each of R and S get wrapped in an integer, which has a 1 byte tag (02) and a variable-byte length (1 for < 0x80, 2 for 0x80-0xFF, 3 for 0xFF-0xFFFF, 4 for 0xFFFF-0xFFFFFF, ...), then those two values get wrapped in a SEQUENCE (1 byte tag, same variable length).  Integers with the high bit set will gain a padding byte to prevent them from being treated as a negative number.

So for secp256r1 the maximum length for the integer contents is `1 + 256/8` = 33 bytes.  33 bytes produce a prefix of 0221.  The maximum given to the sequence is 2x35 bytes == 70 bytes, so it'll start with (maximally) 3046.  6 bytes held.

For secp521r1 a surprise padding byte is never needed (since 521 is not byte aligned).  It needs a maximum of (521+7)/8 bytes (66) for R or S's content, so the prefix is 0242.  But, put two of those together, and you have 68*2 = 136 bytes for the sequence.  That's more than 127, so the prefix is 308188... 3 bytes.  This one took 7.

So 7 is the practical maximum (there aren't really bigger curves yet, because they're 'pointlessly secure').

If you haven't had an assert trip on this yet it's because it requires both R and S to have their highest bit set, which sounds like 25%, but it's less than that since there are fewer legal values with the bit set than without them being set.  I feel like it should come out to around 18% of the time that it needed 7 bytes of overhead, but that's recall off the top of my head, not using math freshly.

jkoritzinsky replied:
Java doesn't supply a method to get the size since they provide an API that returns an array of the right length, so we have to handle the calculation ourselves. Maybe we can calculate the max DER overhead based on the length of the order + the extra bytes for the padding bytes?

bartonjs replied:
Can you use the "inP1363Format" versions described in https://docs.oracle.com/en/java/javase/12/docs/specs/security/standard-names.html#signature-algorithms?  Their documentation is wrong, but hopefully their code is right.

The P1363 format is the .NET default, r and s as a fixed-length unsigned encoding each as long as the size of the order.

Of course, that would mean you have to split off from ECDsaOpenSsl already, since ECDsaOpenSsl expects the X.509 DER version.

jkoritzinsky replied:
Yes, I should be able to do that. I was hoping to avoid having to make changes to managed code in this PR, but I can do it if need be.

jkoritzinsky replied:
Android doesn't support the "inP1363Format" variations, so it looks like we'll have to fix this.

bartonjs replied:
The _easiest_ fix is to set the overhead to 18.  That would be where each of R and S are between 0x0100_0000 and 0x7FFF_FFFE long and the resulting sequence then contains 0xFFFF_FFFE payload bytes (each segment adds a 1 byte tag and 5 byte length, 6*3 = 18).

The hybrid of easy and not wasteful would be to set the overhead to 7, which covers up through secp521r1.  (It's good for until R/S exceed 0xFF bytes, which would be at 255*8-1 (2039) bits... so probably good enough).

For perfect accuracy, you'll want something like https://source.dot.net/#System.Formats.Asn1/System/Formats/Asn1/AsnWriter.cs,aa1dbe1cb6c5929e (though you need to add one for the "non-subsequent" byte that this class already knew about, and one for the tag), and to do 

```
int rsMax = SegmentSize((keySizeInBits + 8) / 8);
int seqMax = SegmentSize(rsMax * 2);
return seqMax;
```

Why +8 / 8?  "Add one bit for the sign bit, then it's bytes, round up".  0-7 bits: 1 byte.  8-15: 2 bytes, etc.

But, that'll just return 6 for KeySize <= (((0x7F / 2) - 2) * 8) - 1 = 487 bits, and 7 for up to 2039 bits.  The biggest defined in OpenSSL is sect571r1, at 571 bits... so 7 is probably safe.

jkoritzinsky replied:
I've set it to 7 for now.

VSadov commented:
Regarding c++ and singlefile - I do not think singlefile linking would care about C vs. C++. 
We have a mix anyways. Some components are definitely C, like some zlib stuff uses K&R syntax and FAR pointers, 
and we have components like fxr resolver, which is using `std::vector<T>` and such.

jkoritzinsky commented:
I think the issue for C vs C++ was size on disk and working set memory.

bartonjs commented:
> AFAIK, System.Native.* libs were initially C++ in corefx 

Initially they were C.  Then there were some nice-to-uses in C++ that caused things to get converted.  Then there were reasons to convert back, that I don't remember.

My recommendation is to stick with C.

jkoritzinsky committed:
Fix crash.

jkoritzinsky committed:
Merge branch 'android-ecc' of github.com:jkoritzinsky/runtime into android-ecc

jkoritzinsky committed:
PR style feedback.

jkoritzinsky committed:
Error out gracefully if private key is requested and only public key is known.

jkoritzinsky committed:
Test for overflow.

jkoritzinsky committed:
Add ReleaseLRef method and use it in the ECC apis.

marek-safar commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,132 @@
+int32_t CryptoNative_EcKeyGetCurveName2(const EC_KEY* key, int32_t* nidName)
+{
+    // TODO: The public APIs do not support getting a name from an ECKeyParameters object.
+    return FAIL;
```
Why not block this in managed code?

jkoritzinsky replied:
One of my goals with this PR was to not change the managed surface area if possible. I was just trying to fill in the shim.

It seems like that might not be possible now, so I'll fold this into the custom managed code for Android.

jkoritzinsky replied:
I've made this safely no-op in the shim for now since the reason to make lots of managed code changes in this PR (using the P1363 encoding for the DSA signatures instead of DER encoding) isn't happening since Android doesn't support the P1363 formats.

am11 commented:
IIRC, the main reason to convert the code C++ to C in CoreFX days was mono's preference for C, at the time mono started to consume CoreFX libraries (for mono/corefx fork).

Additionally, it noticeably reduces the [code size](http://www.on-time.com/ddj0011.htm) of native libraries, when it was compared SxS. Unless we need exception handling or C++ specific features (like corehost currently has this requirement), sticking to plain old C pays off in multiple ways.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,558 @@
+    return rc;
+}
+
+static jobject CryptoNative_CreateKeyPairFromCurveParameters(jobject curveParameters, uint8_t* qx, int32_t qxLength, uint8_t* qy, int32_t qyLength, uint8_t* d, int32_t dLength)
```
At one point we were trying to be good at running `clang-format -i <file>` on the shim files.  I'm pretty sure it will rewrite this into a style I don't personally like, but it does provide some uniformity.

jkoritzinsky committed:
Run clang-format -i on the files.

jkoritzinsky committed:
Bump der encoding extra bytes.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

jkoritzinsky committed:
Provide no-op implementation of EcKeyGetCurveName2

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

jkoritzinsky committed:
ECC lifetime and cleanup fixes.

jkoritzinsky committed:
More exception handling.

jkoritzinsky committed:
Check if nistP384 is supported on the platform before including it. This will enable the ECDsaSignatureFormatTests suite.

jkoritzinsky committed:
Call the correct KeyPairGenerator.initialize override.

jkoritzinsky committed:
Manually convert some well-known curve oids to their friendly names since Android versions we support don't support the oids directly.

jkoritzinsky committed:
Fix some lifetime issues.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

jkoritzinsky committed:
Fix initialization bug

lambdageek commented:
> Well, this lib is a standalone dynamic lib so it doesn't matter at all - cmake will just build a `*.so` and Mono will be able to pinvoke it.
> **UPD** even if it's linked statically - we still use Android NDK's C++ compiler to build the mono runtime as far as I remember.
> cc-ying @akoeplinger @lambdageek to verify my statements regarding C++

The issue is precisely with the NDK - as far as I know, Xamarin.Android builds with the "none" C++ library -  https://developer.android.com/ndk/guides/cpp-support (since "system" is deprecated, and "libc++" means every xamarin app will bundle a copy of the C++ library which is not acceptable in terms of size for the XA team).  So you can, kind of, use C++, as long as oyu don't use any features that will require anything from the runtime library which is a fairly constrained language.

AaronRobinsonMSFT commented:
>So you can, kind of, use C++, as long as you don't use any features that will require anything from the runtime library which is a fairly constrained language.

@lambdageek That was what I was worried about. Thanks for confirming this requirement.

grendello commented:
> > Well, this lib is a standalone dynamic lib so it doesn't matter at all - cmake will just build a `*.so` and Mono will be able to pinvoke it.
> > **UPD** even if it's linked statically - we still use Android NDK's C++ compiler to build the mono runtime as far as I remember.
> > cc-ying @akoeplinger @lambdageek to verify my statements regarding C++
> 
> The issue is precisely with the NDK - as far as I know, Xamarin.Android builds with the "none" C++ library - https://developer.android.com/ndk/guides/cpp-support (since "system" is deprecated, and "libc++" means every xamarin app will bundle a copy of the C++ library which is not acceptable in terms of size for the XA team). So you can, kind of, use C++, as long as you don't use any features that will require anything from the runtime library which is a fairly constrained language.
> 
> /cc @grendello

The size is a concern, yes, but a bigger concern is that if we build with one version of `libc++` and an application brings in another version of it, we have a problem on our hands. The only way we could use `libc++` would be if we were able to rename the standard C++ namespaces so that there's no possible conflict. However, you CAN use stuff from the standard library if you're careful enough - you can use things like `std::array` for instance, anything that's header only should be fine.

jkoritzinsky committed:
Get some ECDSA tests passing.

jkoritzinsky committed:
Split off test into two tests since Android doesn't support getting the name of an already instantiated curve object.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

jkoritzinsky committed:
Enable generating keys when we have an explicit curve but not explicit keys.

jkoritzinsky committed:
All bignums should be created as positive values.

All of the algorithms we support that use BigIntegers operate in modulus fields, where all negative values are congruent to a positive value. As a result, we don't use the negative values in the crypto implementations.

For example, explicitly specifying the NistP256 curve has a P, A, and Order that all start with FF in the byte array, but should be positive values.

jkoritzinsky committed:
Get a lot more ECC tests passing.

marek-safar commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs:
```diff
@@ -58,7 +58,8 @@ private static byte[] TryWithOutputArray(Func<byte[], (bool, int)> func)
     {
         [Fact]
-        public void KeySizeProp()
+        [SkipOnMono("Android doesn't support getting a curve name from a curve.", TestPlatforms.Android)]
```
Could you please annotate not supported curve name APIs with `UnsupportedOSPlatformAttribute("android")]`

jkoritzinsky replied:
I'd rather mark the APIs when we get closer to finishing the Android crypto work in case we figure out a way to support these APIs instead of marking them now and then having to figure out which ones to un-mark later.

bartonjs replied:
Quick searching suggests that `PrivateKey::getEncoded()` returns the PKCS#8 blob.  That probably includes the encoded curve and further probably includes it as a named curve when the named curve is known.

jkoritzinsky replied:
I found a way to do this on Android via an API exposed via reflection, so the API is no longer unsupported.

jkoritzinsky committed:
Use the correct variant of ECDSA. We need to use the NONE variant since we do the hashing in managed code.

jkoritzinsky committed:
Be more explicit about leading-trailing zeros.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

jkoritzinsky commented:
This PR is passing a large majority of the ECC and ECDSA tests, with most failing due to the lack of OID mapping support currently in the shim. I believe this is ready for further review.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -7,6 +7,26 @@
+    if (data[0] & 0x80)
+    {
+        // If the top bit is set, we need to create a BigInteger with the sign-and-magnitude constructor.
+        bigNum = (*env)->NewObject(env, g_bigNumClass, g_bigNumCtorWithSign, 1, buffArray);
```
My initial reaction was "why is this working for OpenSSL, then?", but:

> BN_bin2bn() converts the positive integer in big-endian form of length len at s into a BIGNUM and places it in ret. If ret is NULL, a new BIGNUM is created.

So the most correct action would be to fix the existing routines to always do a positive import.

Similarly, the data export is

> BN_bn2bin() converts the absolute value of a into big-endian form and stores it at to. to must point to BN_num_bytes(a) bytes of memory.


Being consistent with those seems easier while you're in the hybrid state.

bartonjs replied:
Actually, looks like you did change that one to always be unsigned.  So how is this version different (besides being longer, since I think both the if and else could both call the if)?

jkoritzinsky replied:
This was my first attempt to work around the leading-zero-byte issue. I'll go back to reusing the bignum PAL apis.

jkoritzinsky committed:
Fix broken merge.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_bignum.c:
```diff
@@ -20,12 +20,41 @@ int32_t CryptoNative_BigNumToBinary(jobject bignum, uint8_t* output)
+
+    // We strip the leading zero byte from the byte array.
+    jbyte leadingByte;
+    (*env)->GetByteArrayRegion(env, bytes, 0, 1, &leadingByte);
```
Possibly cheaper to use the `signum()` function, and subtract one if it returns a negative number?

bartonjs replied:
https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#signum()

jkoritzinsky replied:
The issue here is that the BigInteger is positive due to the leading zero byte, so the signum function here isn't helpful.

bartonjs replied:
Ah, right, it's the positive numbers where the high bit in the high byte is set.  Not sure what negative numbers I was thinking of.

Still possibly more performant than the toByteArray method is using bitLength (if it's a multiple of 8 and signum is >0 then you'll get the padding byte). At least, assuming I read the docs correctly :smile:.

jkoritzinsky replied:
I'll test out some cases and see if I can do this more efficiently.

jkoritzinsky replied:
I tested it out and this won't work. `bitlength` doesn't count the sign bit even when the BigInteger is `0xffffff` and forced positive by the sign+magnitude constructor.

I think we have to do it the way I'm doing it now.

bartonjs replied:
> bitlength doesn't count the sign bit

Yeah, that matches the docs: "Returns the number of bits in the minimal two's-complement representation of this BigInteger, excluding a sign bit.".

I think it's something like

```
int bitLen = bigInt.bitLength();
return (bitLen + 7) / 8;
```

(Might be wrong for negative numbers if it really "excludes" the sign bit, vs excludes the cleared sign bit for positive numbers... then `if (bigInt.signum() < 0) { bitLen++; }`)

My expectation from the docs is that positive 0xFF will report signum() as 1 and bitLength() as 8, but export in two bytes.

(8 + 7) / 8 == 15 / 8 == 1

The extractor wouldn't need to change.  This just detects the same thing without going through array export.

jkoritzinsky replied:
Yes, your expectation is correct.

I'll update this to use the `(bitlength+7) / 8` calculation for the case that strips the leading zero byte.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,641 @@
+    JNIEnv* env = GetJNIEnv();
+
+    // Get the public key
+    jobject publicKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPublicMethod);
```
I feel like there are a lot more `ON_EXCEPTION_GOTO(fail)`s (or whatever the right words are :smile:) needed, since anything that allocates a return could OOM.

jkoritzinsky replied:
We've decided that since OOM is already a fatal exception, once we hit OOM we're probably going down anyway. As a result, we don't need to check exceptions after every call since JNI will take down the process if there's an exception in progress when we call another JNI method.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,641 @@
+    if (cbD)
+        *cbD = 0;
+    ReleaseLRef(env, xBn);
+    ReleaseLRef(env, yBn);
```
Might be worth using @elinor-fung's VA_ARG pattern from the X509 version to make LRef cleanup more consistent.

jkoritzinsky replied:
I'm personally not a fan of that pattern and prefer the explicit calls here. Especially since the ECC code has to use different cleanup methods depending on where the object was created from.

AaronRobinsonMSFT replied:
Not going to push this but I would also recommend a similar mechanism. When there are this many locals that need management it is much harder to update and make sure one isn't missed.

jkoritzinsky replied:
I'll give it another look then and see if I can make it work for the ECC code.

jkoritzinsky committed:
Use bignum apis since they handle the sign/leading zero issues well enough now.

jkoritzinsky committed:
Split Android ECC impl and modify the pal APIs to enable getting a curve name from a path from a non-public API exposed via reflection.

Add curve name OID mappings for the Java names for some of the curves.

jkoritzinsky commented:
I've split the Android and OpenSSL managed code shims to enable getting the curve name from curve parameters and removing some no-op APIs.

This also required making a copy of the ECDiffieHellman support from OpenSSL into Android, but that API still needs work to get fully functional.

jkoritzinsky committed:
Implement ECDH on Android.

jkoritzinsky commented:
Since I was already in the metaphorical neighborhood, I've also implemented the ECDH key agreement algorithm on Android instead of leaving it in a no-op state.

There's no other big ECC chunks in the Algorithms assembly, so this PR is feature-complete barring review feedback.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+
+error:
+    ReleaseLRef(env, signatureObject);
+    return -1;
```
How is this different from `FAIL` above?

jkoritzinsky replied:
Like RSA, some of these functions on OpenSSL return 1 for success, 0 for completion but failure (like a failed signature verification), and -1 for error.

I’ve adopted the same convention here for simplicity since the managed side still expects the same codes.

AaronRobinsonMSFT commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c:
```diff
@@ -0,0 +1,114 @@
+
+int32_t CryptoNative_EcDsaSign(const uint8_t* dgst, int32_t dgstlen, uint8_t* sig, int32_t* siglen, EC_KEY* key)
+{
+    if (!siglen)
```
So the rest of the arguments don't need to be checked?

jkoritzinsky replied:
I've added some asserts for the other inputs.

AaronRobinsonMSFT reviewed (approved):
There is a lot going on here and if @bartonjs is okay with this then I am okay since all this code is internal and mostly glue.

jkoritzinsky committed:
Use the INIT_LOCALS pattern to assist in handling lifetimes in ECC import/export.

jkoritzinsky committed:
Add some asserts for ECDSA sign/verify paramters that will never be null.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

elinor-fung commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECAndroid.ImportExport.cs:
```diff
@@ -0,0 +1,191 @@
+{
+    internal sealed partial class ECAndroid
+    {
+        internal const string ECDSA_P256_OID_VALUE = "1.2.840.10045.3.1.7"; // Also called nistP256 or secP256r1
+        internal const string ECDSA_P384_OID_VALUE = "1.3.132.0.34"; // Also called nistP384 or secP384r1
+        internal const string ECDSA_P521_OID_VALUE = "1.3.132.0.35"; // Also called nistP521or secP521r1
```
Can we not use the constants in `Oids.cs`?

elinor-fung commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDsaAndroid.cs:
```diff
@@ -0,0 +1,376 @@
+
+namespace System.Security.Cryptography
+{
+        public partial class ECDsa : AsymmetricAlgorithm
```
Alignment.

elinor-fung commented on src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.EcDsa.ImportExport.cs:
```diff
@@ -0,0 +1,334 @@
+            if (rc == -1)
+            {
+                key?.Dispose();
+                Interop.Crypto.ErrClearError();
```
I'd remove all the `ErrClearError` calls. The implementation in AndroidCrypto is empty (we should just remove it as we keep splitting off).

elinor-fung commented on src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.EcDsa.ImportExport.cs:
```diff
@@ -0,0 +1,334 @@
+            {
+                if (key != null)
+                    key.Dispose();
+                throw Interop.Crypto.CreateOpenSslCryptographicException();
```
I think we just want a normal `CryptographicException` instead of `OpenSslCryptographicException`.

elinor-fung replied:
Actually, I think this will actually throw a `CryptographicException` with AndroidCrypto - but with an extra P/Invoke to get there.

elinor-fung reviewed (commented):
Very much glazed over the actual algorithm/logic part of this - I'm happy as long as the experts in this area are happy.

jkoritzinsky committed:
Remove old logging statements

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_jni.c:
```diff
@@ -61,7 +61,11 @@ jmethodID g_ivPsCtor;
 jmethodID g_toByteArrayMethod;
+jmethodID g_valueOfMethod;
+jmethodID g_intValueMethod;
+jmethodID g_compareToMethod;
```
I was following the convention for the BigInteger method handles. I can convert these if we want to.

jkoritzinsky commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,201 @@
+    jobject paramSpec = (*env)->NewObject(env, g_ECGenParameterSpecClass, g_ECGenParameterSpecCtor, oidStr);
+    ReleaseLRef(env, oidStr);
+
+    jobject keyPairGenerator =
```
At least for me, my brain has already started re-parsing the JNI as though it's Java. I'm not the biggest fan of the comments since they're can easily get out of date and are sometimes just a simplification of what's happening (and not a direct 1:1).

jkoritzinsky committed:
Rename android apis to be android-specific. Remove OpenSSL-style error handling.

jkoritzinsky committed:
Use Oids constants.

jkoritzinsky committed:
xBn and yBn have already been released by the time we can jump to the error label.

jkoritzinsky commented:
@bartonjs any more comments from the crypto side of things?

jkoritzinsky committed:
Keep constants for OpenSSL EC implementation due to layering issues.

jkoritzinsky committed:
Use bitLength method to speed up GetBigNumBytes.

bartonjs commented:
> @bartonjs any more comments from the crypto side of things?

If it's otherwise ready, I'll get back to this tomorrow morning (Redmond time).

bartonjs commented on src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.EcKey.cs:
```diff
@@ -0,0 +1,76 @@
+                }
+
+                string curveNameStr = Marshal.PtrToStringUni(curveName)!;
+                Marshal.ZeroFreeCoTaskMemUnicode(curveName);
```
The name of a curve isn't secret, I'd just go with Marshal.FreeHGlobal.

bartonjs commented on src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Dsa.cs:
```diff
@@ -168,5 +168,40 @@ internal static DSAParameters ExportDsaParameters(SafeDsaHandle key, bool includ
+        /// <summary>
+        /// Return the maximum value in the array; assumes non-negative values.
+        /// </summary>
+        private static int GetMax(int[] values)
```
These don't seem particularly relevant to DSA...

jkoritzinsky replied:
They’re used by the interop code for DSA and ECDSA. When I split the ECDSA code, I needed these methods to live somewhere bin the Android build since the Android ECDSA interop is in the AndroidCrypto class now.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,572 @@
+    ReleaseGRef(env, *seed);
+    *p = *a = *b = *gx = *gy = *order = *cofactor = *seed = NULL;
+
+    RELEASE_LOCALS_ENV(loc, ReleaseLRef);
+    RELEASE_LOCALS_ENV(bn, ReleaseLRef);
+
+exit:
+    return rc;
```
Shouldn't the locals be released in the success path, too?

jkoritzinsky replied:
They’re released when I call ToGRef to convert them to global references.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,572 @@
+// Returns -1 if bigInteger < 0 or > INT32_MAX
+static int32_t ConvertBigIntegerToPositiveInt32(JNIEnv* env, jobject bigInteger)
+{
+    jobject zero = (*env)->CallStaticObjectMethod(env, g_bigNumClass, g_valueOfMethod, (int64_t)0);
+    int isPositive = (*env)->CallIntMethod(env, bigInteger, g_compareToMethod, zero);
+    (*env)->DeleteLocalRef(env, zero);
```
Seems like `signum()` (-1, 0, or 1) would work here and involve less handshaking with JNI

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecc_import_export.c:
```diff
@@ -0,0 +1,572 @@
+        return -1;
+    }
+
+    jobject int32MaxBigInt = (*env)->CallStaticObjectMethod(env, g_bigNumClass, g_valueOfMethod, (int64_t)INT32_MAX);
+    int willOverflow = (*env)->CallIntMethod(env, bigInteger, g_compareToMethod, int32MaxBigInt);
+    (*env)->DeleteLocalRef(env, int32MaxBigInt);
```
In a `.java` file I'd suggest promoting this to a static field to reduce the GC cost of creating and destroying repeated copies of an immutable value.

If there's an easy way to get similar semantics from JNI without being afraid of lifetime semantics, maybe it makes sense to do so.  If you hesitate at all, it might not be worth it.

jkoritzinsky replied:
It'll be a bit of work to get similar semantics with thread safety, so I'd rather leave it as it is at least for now.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,193 @@
+    if (r)
+    {
+        r->refCount--;
+        if (r->refCount == 0)
```
We probably want to care about thread safety for refcounting, especially depending on how things like cert.GetECDsaPrivateKey() get implemented (if 20 different managed objects are all tracking that they're backed by the same `EC_KEY*` then there's no way the callers would know that their 20 parallel Dispose() calls aren't thread-kosher)

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,193 @@
+
+    jobject curveParameters = (*env)->CallObjectMethod(env, keySpec, g_ECPublicKeySpecGetParams);
+
+    jobject group = (*env)->CallObjectMethod(env, curveParameters, g_ECParameterSpecGetCurve);
+
+    jobject field = (*env)->CallObjectMethod(env, group, g_EllipticCurveGetField);
+
+    ReleaseLRef(env, field);
+    ReleaseLRef(env, group);
```
Are these lines still needed?  If there's a side effect from calling them, please add a comment so that the code doesn't get overzealously cleaned up.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,193 @@
+
+    JNIEnv* env = GetJNIEnv();
+
+    jobject curve = (*env)->CallObjectMethod(env, key->curveParameters, g_ECParameterSpecGetCurve);
+    jobject field = (*env)->CallObjectMethod(env, curve, g_EllipticCurveGetField);
+    *keySize = (*env)->CallIntMethod(env, field, g_ECFieldGetFieldSize);
```
I don't remember how often we ask the native code for the keysize (vs trust that we loaded it correctly into the KeySize property).  If it's a lot, maybe this is something that we'd want to calculate once then pass in to NewEcKey (or calculate once inside NewEcKey).  If it's only once or so, the compartmentalization here is probably better.

jkoritzinsky replied:
I think we most often just use the KeySize property.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_jni.h:
```diff
@@ -236,10 +323,12 @@ void SaveTo(uint8_t* src, uint8_t** dst, size_t len, bool overwrite);
 bool CheckJNIExceptions(JNIEnv* env);
 void AssertOnJNIExceptions(JNIEnv* env);
 jmethodID GetMethod(JNIEnv *env, bool isStatic, jclass klass, const char* name, const char* sig);
+jmethodID TryGetMethod(JNIEnv *env, bool isStatic, jclass klass, const char* name, const char* sig);
```
In .NET Try parlance this isn't really Try (doesn't return a boolean).  GetOptionalMethod, perhaps?

If it's a reasonable "Try" from common JNI perspective, then OK.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECAndroid.ImportExport.cs:
```diff
@@ -0,0 +1,183 @@
+        public static ECParameters ExportParameters(SafeEcKeyHandle currentKey, bool includePrivateParameters)
+        {
+            ECParameters ecparams;
+            if (Interop.AndroidCrypto.EcKeyHasCurveName(currentKey))
```
This is going to ask for the name, see it's not null, throw away the string, then ask for it again.

Consider changing this to a flow where you ask for the name, and if it's non-null pass it into ExportNamedCurveParameters, otherwise you call ExportExplicit.

jkoritzinsky committed:
PR Feedback.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into android-ecc

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellmanAndroid.Derive.cs:
```diff
@@ -0,0 +1,196 @@
+                    // Arbitrary limit. But it covers secp521r1, which is the biggest common case.
+                    const int StackAllocMax = 66;
+
+                    if (secretLength > StackAllocMax)
```
Wasn't secretLength set to KeySize, so it's in bits whereas the StackAllocMax is in bytes?

jkoritzinsky replied:
Yes. This looks to also be a bug in the OpenSSL implementation. I'll fix both.

jkoritzinsky replied:
Actually nevermind the OpenSSL impl asks OpenSSL for the size.

jkoritzinsky committed:
Convert KeySize to bytes for secretlength.

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellmanAndroid.Derive.cs:
```diff
@@ -0,0 +1,196 @@
+                        secret = stackalloc byte[secretLength];
+                    }
+
+                    if (!Interop.AndroidCrypto.EcdhDeriveKey(ourKey, theirKey, secret, out int usedBufferLength))
+                    {
+                        throw new CryptographicException();
+                    }
+
+                    secret = secret.Slice(0, usedBufferLength);
```
Presumably we want to assert that usedBufferLength was the expected value?

If it was shorter than expected then we might have to manually insert some padding zeros... but I'm hoping that the ECDH implementation did that for us.  (I think it's a 50/50 to get a leading 0x00 in secp521r1 ECDH, so shouldn't take long to verify)

jkoritzinsky replied:
Should I just run ECDH a few times to see if I get a leading zero with that curve?

bartonjs replied:
Yeah.  Make sure that at least one of the two keys varies, though.  Otherwise it'll just produce the same answer over and over again :smile:.

jkoritzinsky replied:
Ran this 10 times with randomized keys and never saw a mismatch between `secretLength` and `usedBufferLength`. Added an assert to continue to validate that they match.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,189 @@
+{
+    if (r)
+    {
+        int count = --r->refCount;
```
Should be atomic

jkoritzinsky replied:
This is atomic. The type of `refCount` is a C atomic int.

bartonjs commented on src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_eckey.c:
```diff
@@ -0,0 +1,189 @@
+{
+    if (!r)
+        return FAIL;
+    r->refCount++;
```
Should be atomic

jkoritzinsky replied:
This is atomic. The type of `refCount` is a C atomic int.

bartonjs replied:
D'oh! 🤦 

I just searched for "refcount", saw one line had the word atomic and the other two didn't.  I thought that each atomic increment/decrement looked something like `atomic_increment(&key->refCount, 1, ATOMIC_ALL_THE_THINGS)`

bartonjs commented on src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/ECDiffieHellmanAndroid.Derive.cs:
```diff
@@ -0,0 +1,197 @@
+                SafeEcKeyHandle? theirKey = null;
+                byte[]? rented = null;
+                // Calculate secretLength in bytes.
+                int secretLength = (KeySize + 7) / 8;
```
For all the places in managed code that do this (F3 shows me two instances of "+ 7" in this PR, but some might be folded away), I got feedback before that this formula (burned into my brain) requires too much thinking.  So there's a helper method for it: AsymmetricAlgorithmHelpers.BitsToBytes(bitLength) => byteLength.

jkoritzinsky replied:
One occurrence I believe is in native code, so I'll leave that one as is. I'll change the other one to use the helper.

bartonjs reviewed (approved):
Two outstanding bits of housekeeping:

* Use BitsToBytes instead of repeating the formula in managed code
* After verifying how leading zeros work, either assert or handle the case where bytesWritten in ECDH isn't the expected value.

bartonjs commented on src/libraries/Common/tests/System/Security/Cryptography/AlgorithmImplementations/ECDsa/ECDsaTests.netcoreapp.cs:
```diff
@@ -227,6 +227,7 @@ public void TestRegenKeyNistP256()
 
         [Theory]
         [MemberData(nameof(TestCurves))]
+        [SkipOnMono("Checking if a curve is named is unsupported on Android", TestPlatforms.Android)]
```
Didn't you get this working?

jkoritzinsky replied:
I did. I just forgot to re-enable the test.

jkoritzinsky committed:
Reenable test and add assert.

jkoritzinsky committed:
Merge branch 'master' into android-ecc

jkoritzinsky committed:
Add missing using.

jkoritzinsky committed:
Merge branch 'android-ecc' of github.com:jkoritzinsky/runtime into android-ecc

jkoritzinsky commented:
Merging in as this PR has no effect on iOS or the standard Android build with OpenSSL.
